### DISCUSSION TITLE
What's a good way to increase mask recognition accuracy?

### DISCUSSION THREAD

**Discussion by nadongjin** [leondgarse/Keras_insightface#64] [Q&A]
I ask you for advice.
What is a good way to increase mask recognition accuracy?
Currently, I am going to choose to draw a mask using face landmark, but the problem is that the data set doubles.
Another method is to draw a mask only on the front face in the dataset, and the side face tries to choose the flip method, but the data set also increases too much.
Is there a good way?

**Comment by leondgarse** [comment]
I think you may randomly apply masks during training, something like:
```py
def apply_face_mask_to_image(image):
    # Apply face mask
    masked_face = image
    return masked_face

def random_apply_face_mask(image, label, apply_ratio=0.5):
    image = tf.cond(
        tf.random.uniform(()) < apply_ratio,
        lambda : apply_face_mask_to_image(image),
        lambda : image
    )
    return image, label
```
Then call `ds = ds.map(random_apply_face_mask, num_parallel_calls=AUTOTUNE)` before `ds = ds.batch`.
Something related is the best face mask detector I tried comes form [Github neuralet/models](https://github.com/neuralet/models). Their article: [A practical solution to real-world Face Mask Detection – Part 1](https://neuralet.com/article/real-time-face-mask-detection-part-one).

**Reply by nadongjin** [reply]
very thanks..
https://github.com/aqeelanwar/MaskTheFace
this site good for draw face mask...

**Reply by leondgarse** [reply]
Just an option you may consider. IMO, those complicated face masks are essential for mask detector, but may not that necessary for face recognition. Thus maybe just ignoring that area by some ratio can also work. I just add an option param for `train.Train` `random_cutout_mask_area` in [Add random_cutout_mask_area params](https://github.com/leondgarse/Keras_insightface/commit/1e8b9554c8129f24bece72aeb8905fea0948d44e). It will set a ratio for randomly cutout bottom **2/5** area, regarding as ignoring mask area. This operation is applied on a whole batch.
```py
import data
ds, steps_per_epoch = data.prepare_dataset('/datasets/faces_casia_112x112_folders/', batch_size=64, random_cutout_mask_area=0.5)
data.show_batch_sample(ds)
```
![cutout_face_mask](https://user-images.githubusercontent.com/5744524/147866113-7c5bd8ff-6350-4a41-bf23-3c5408219937.png)

**Reply by nadongjin** [reply]
I always appreciate your advice.
Thank you.
